pkg/service/controller/servingcert/controller.go

```diff
@@ -0,0 +1,291 @@
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+
+	sc.queue.Add(key)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (sc *ServiceServingCertController) worker() {
+	for {
+		if !sc.worker_inner() {
+			return
+		}
+	}
+}
+
+// worker_inner returns true if the worker thread should continue
+func (sc *ServiceServingCertController) worker_inner() bool {
+	key, quit := sc.queue.Get()
+	if quit {
+		return false
+	}
+	defer sc.queue.Done(key)
+
+	if err := sc.syncHandler(key.(string)); err == nil {
+		// this means the request was successfully handled.  We should "forget" the item so that any retry
+		// later on is reset
+		sc.queue.Forget(key)
+
+	} else {
+		// if we had an error it means that we didn't handle it, which means that we want to requeue the work
+		utilruntime.HandleError(fmt.Errorf("error syncing service, it will be retried: %v", err))
+		sc.queue.AddRateLimited(key)
+	}
+
+	return true
+}
+
+// syncService will sync the service with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (sc *ServiceServingCertController) syncService(key string) error {
+	obj, exists, err := sc.serviceCache.GetByKey(key)
+	if err != nil {
+		glog.V(4).Infof("Unable to retrieve service %v from store: %v", key, err)
+		return err
+	}
+	if !exists {
+		glog.V(4).Infof("Service has been deleted %v", key)
+		return nil
+	}
+
+	// make a copy to avoid mutating cache state
+	t := *obj.(*kapi.Service)
+	service := &t
+
+	if !sc.requiresCertGeneration(service) {
+		return nil
+	}
+
+	dnsName := service.Name + "." + service.Namespace + ".svc"
+	fqDNSName := dnsName + "." + sc.dnsSuffix
+	servingCert, err := sc.ca.MakeServerCert(sets.NewString(dnsName, fqDNSName))
+	if err != nil {
+		return err
+	}
+	certBytes, keyBytes, err := servingCert.GetPEMBytes()
+	if err != nil {
+		return err
+	}
+
+	secret := &kapi.Secret{
+		ObjectMeta: kapi.ObjectMeta{
+			Namespace: service.Namespace,
+			Name:      service.Annotations[ServingCertSecretAnnotation],
+			Annotations: map[string]string{
+				ServiceUIDAnnotation:  string(service.UID),
+				ServiceNameAnnotation: service.Name,
+			},
+		},
+		Type: kapi.SecretTypeTLS,
+		Data: map[string][]byte{
+			kapi.TLSCertKey:       certBytes,
+			kapi.TLSPrivateKeyKey: keyBytes,
+		},
+	}
+
+	_, err = sc.secretClient.Secrets(service.Namespace).Create(secret)
+	if err != nil && !kapierrors.IsAlreadyExists(err) {
+		// if we have an error creating the secret, then try to update the service with that information.  If it fails,
+		// then we'll just try again later on  re-list or because the service had already been updated and we'll get triggered again.
+		service.Annotations[ServingCertErrorAnnotation] = err.Error()
```

Comment:
> I never thought about the other direction. Did you ever see it working like that?

I was pretty sure that our fuzzer behavior got weird because of it.